main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
+        let Iterator(_, iterF) = iter;
+        def loop() = match iterF() {
+            case Ans(a) => Some(a)
+            case Skip   => loop()
+            case Done   => None
+        };
+        loop()
 
     ///
     /// Returns an iterator of all integers between `b` (inclusive) and `e` (exclusive).
     ///
     /// Returns an empty iterator if `b >= e`.
     ///
-    @Lazy
-    pub def range(r: Region[r], b: Int32, e: Int32): Iterator[Int32, r] \ Write(r) =
-        let i = ref b @ r;
-        let done = () -> deref i >= e;
-        let next = () -> {
-                let res = deref i;
-                if (res < e) {
-                    i := res + 1;
-                    res
-                }
-                else {
-                    bug!("Empty Iterator!")
-                }
+    pub def range(rc: Region[r], b: Int32, e: Int32): Iterator[Int32, ef, r] \ Write(r) =
+        let pos = ref b @ rc;
+        if (e <= b)
+            empty(rc)
+        else {
+            let f = () -> match (deref pos) {
+                case i if i < e => pos := i+1; Ans(i)
+                case _          => Done
             };
-        Iterator(done, next)
+            let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+            Iterator(rc, iterF)
+        }
 
     ///
     /// Returns an iterator over an iterable with the element `x` repeated `n` times.
     ///
     /// Returns an empty iterator if `n < 0`.
     ///
-    @Lazy
-    pub def repeat(r: Region[r], n: Int32, x: a): Iterator[a, r] \ Write(r) =
-        let i = ref 0 @ r;
-        let done = () -> deref i >= n;
-        let next = () -> {
-              let j = deref i;
-              if (j < n) {
-                i := j + 1;
-                x
-              }
-              else {
-                bug!("Empty Iterator!")
-              }
-            };
-        Iterator(done, next)
+    pub def repeat(rc: Region[r], n: Int32, x: a): Iterator[a, ef, r] \ Write(r) =
+        let ix = ref n @ rc;
+        let f = () -> {
+            let i = deref ix;
+            if (i < 1)
+                Done
+            else {
+                ix := i-1;
+                Ans(x)
+            }
+        };
+        let iterF = () -> unsafe_cast f() as _ \ { ef, r };
+        Iterator(rc, iterF)
 
     ///
     /// Returns the sum of all elements in the iterator `iter`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sum(iter: Iterator[Int32, r]): Int32 \ Read(r) =
+    pub def sum(iter: Iterator[Int32, ef, r]): Int32 \ { ef, Read(r) } =
         foldLeft((acc, x) -> acc + x, 0, iter)
 
     ///
     /// Returns the sum of all elements in the iterator `iter` according to the function `f`.
     ///
     /// Consumes the entire iterator.
     ///
-    pub def sumWith(f: a -> Int32 \ ef, iter: Iterator[a, r]): Int32 \ { ef, Read(r) } =
+    pub def sumWith(f: a -> Int32 \ ef1, iter: Iterator[a, ef, r]): Int32 \ { ef, ef1, Read(r) } =
```

Comment:
I think in all other code we have been using `ef1` and `ef2` in these situations (instead of `ef` and `ef1`), but maybe you want to keep the associated of `ef` with the iterator? I don't have any opinion, just pointing it out.

Reply:
I missed this - I'll take another look at it.

Reply:
I've changed to use ef1 and ef2 - will be in the next commit.